rust/src/operations/optimize.rs

```diff
@@ -498,11 +502,77 @@ impl MergePlan {
             .collect_vec();
 
         // Interleave the batches
-        let out_batches = util::interleave_batches(batches, Arc::new(indices), false).await?;
+        Ok(
+            util::interleave_batches(batches, indices, 10_000, use_threads)
+                .await
+                .map_err(|err| ParquetError::General(format!("Failed to reorder data: {:?}", err)))
+                .boxed(),
+        )
+    }
+
+    /// Datafusion-based z-order read.
+    #[cfg(feature = "datafusion")]
+    async fn read_zorder(
+        columns: Arc<Vec<String>>,
+        files: MergeBin,
+        object_store: ObjectStoreRef,
+        _use_threads: bool,
+    ) -> Result<BoxStream<'static, Result<RecordBatch, ParquetError>>, DeltaTableError> {
+        use datafusion::execution::memory_pool::FairSpillPool;
+        use datafusion::execution::runtime_env::{RuntimeConfig, RuntimeEnv};
+        // TODO: usethreads
+        use datafusion::prelude::{col, ParquetReadOptions, SessionConfig};
+
+        // TODO: make this configurable.
+        // TODO: push this up and share between bins or maybe an overall runtime.
+        let memory_pool = FairSpillPool::new(8 * 1024 * 1024 * 1024);
+        let config = RuntimeConfig::new().with_memory_pool(Arc::new(memory_pool));
+        let runtime = Arc::new(RuntimeEnv::new(config)?);
+        runtime.register_object_store(&Url::parse("delta-rs://").unwrap(), object_store);
```

Comment:
I'm a little stumped here. Where should this configuration live?

It makes sense to have the memory pool as a global config, IMO. But the object store should be local to here, since it's specific to this particular table's root.

Reply:
Had some similar question a little while back. For a while I thought the `DeltaOps` struct could be a place where session state is stored, but the way we use that right now, it's even shorter lived then the `DeltaTable`.

I guess the only thing that "survives" is the `DeltaTableState`. Maybe a way to go is to have that as an optional field on the`DeltaTable`. Then once the first operation is performed, we populate that field in the returned table and re-use if for subsequent operations. This may also give us a way to capture user configuration for such things that would take a similar route through the code as the storage options do right now.